include/revng/ADT/STLExtras.h

```diff
@@ -53,3 +53,54 @@ concept IsNotTupleEnd
 static_assert(HasTupleSize<std::tuple<>>);
 static_assert(!HasTupleSize<std::vector<int>>);
 static_assert(!HasTupleSize<int>);
+
+//===----------------------------------------------------------------------===//
+//     Extra additions to <iterator>
+//===----------------------------------------------------------------------===//
+
+namespace revng {
+namespace detail {
+template<typename ItTy,
+         typename FuncTy,
+         typename FuncReturnTy = decltype(
+           std::declval<FuncTy>()(*std::declval<ItTy>()))>
+class proxy_mapped_iterator_impl : public llvm::mapped_iterator<ItTy, FuncTy> {
+  struct IteratorProxy {
+    IteratorProxy(FuncReturnTy &&Value) : Temporary(std::move(Value)) {}
+    FuncReturnTy *const operator->() { return &Temporary; }
+    FuncReturnTy const *const operator->() const { return &Temporary; }
+  private:
+    FuncReturnTy Temporary;
+  };
+
+public:
+  using llvm::mapped_iterator<ItTy, FuncTy>::mapped_iterator;
+  IteratorProxy operator->() {
+    return llvm::mapped_iterator<ItTy, FuncTy>::operator*();
+  }
+  IteratorProxy const operator->() const {
+    return llvm::mapped_iterator<ItTy, FuncTy>::operator*();
+  }
+};
+} // namespace detail
+
+// `mapped_iterator` - this is a specialized version of `llvm::mapped_iterator`
+// with improved temporary object support.
+template<typename ItTy, typename FuncTy>
+using mapped_iterator = std::conditional_t<
+  std::is_object_v<decltype(std::declval<FuncTy>()(*std::declval<ItTy>()))>,
+  detail::proxy_mapped_iterator_impl<ItTy, FuncTy>,
+  llvm::mapped_iterator<ItTy, FuncTy>>;
+
+// `map_iterator` - Provide a convenient way to create `mapped_iterator`s,
+// just like `make_pair` is useful for creating pairs...
+template<class ItTy, class FuncTy>
+inline auto map_iterator(ItTy I, FuncTy F) {
+  return mapped_iterator<ItTy, FuncTy>(std::move(I), std::move(F));
+};
+
+template<class ContainerTy, class FuncTy>
+auto map_range(ContainerTy &&C, FuncTy F) {
+  return llvm::make_range(map_iterator(C.begin(), F), map_iterator(C.end(), F));
+}
+} // namespace revng
```

Comment:
This deserves its own header.

Reply:
You think so? I just mirrored the way `llvm` placed this. Since it's not really a new implementation - more like a specialization of their stuff.

Reply:
Yes, but AFAIK overcomes a limitation of `llvm::mapped_iterator` that has been annoying for multiple people.
I think it makes sense to expose it in a header unrelated to `GenericGraph`.

Reply:
Wait, it's not related to `GenericGraph`. This is about `revng/ADT/STLExtras.h` header.

Reply:
Uh, my bad.